http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/KubernetesServiceDbQueries.scala

```diff
@@ -0,0 +1,256 @@
+            case KubernetesClusterStatus.Provisioning => failure
+            case _                                    => DBIO.successful(SaveClusterResult(cluster, true))
+          }
+        case _ => kubernetesClusterQuery.save(saveKubernetesCluster).map(c => SaveClusterResult(c, false))
+      }
+    } yield eitherClusterOrError
+
+  def getActiveFullAppByName(googleProject: GoogleProject, appName: AppName, labelFilter: LabelMap = Map())(
+    implicit ec: ExecutionContext
+  ): DBIO[Option[GetAppResult]] =
+    joinFullAppAndUnmarshal(listByProject(Some(googleProject)),
+                            nodepoolQuery.filter(_.destroyedDate === dummyDate),
+                            appQuery.findActiveByNameQuery(appName),
+                            labelFilter)
+      .map(clusters =>
+        clusters.headOption.fold[Option[GetAppResult]](
+          None
+        )(cluster =>
+          Some(
+            GetAppResult(
+              //we are guaranteed exactly one app and nodepool here if the query doesn't return None due to the way the query JOINs and database unique enforcement
+              cluster,
+              cluster.nodepools.head,
+              cluster.nodepools.head.apps.head
+            )
+          )
+        )
+      )
+
+  //if there are no apps associated with this cluster, None will be returned even if the cluster exists.
+  //see the `Minimal` methods if you wish to get the cluster itself
+  def getFullClusterById(id: KubernetesClusterLeoId)(implicit ec: ExecutionContext): DBIO[Option[KubernetesCluster]] =
+    joinFullAppAndUnmarshal(kubernetesClusterQuery.findByIdQuery(id), nodepoolQuery, appQuery)
+      .map(_.headOption)
+
+  def isDiskAttached(disk: PersistentDisk)(implicit ec: ExecutionContext): DBIO[Boolean] =
+    appQuery.filter(x => x.diskId.isDefined && x.diskId === disk.id).length.result.map(_ > 0)
+
+  private[db] def listByProject(
+    googleProject: Option[GoogleProject],
+    includeDeleted: Boolean = false
+  ): Query[KubernetesClusterTable, KubernetesClusterRecord, Seq] = {
+    val initialQuery =
+      googleProject match {
+        case Some(project) => kubernetesClusterQuery.filter(_.googleProject === project)
+        case None          => kubernetesClusterQuery
+      }
+
+    includeDeleted match {
+      case false => initialQuery.filter(_.destroyedDate === dummyDate)
+      case true  => initialQuery
+    }
+  }
+
+  private[db] def joinFullAppAndUnmarshal(
+    baseQuery: Query[KubernetesClusterTable, KubernetesClusterRecord, Seq],
+    nodepoolQuery: Query[NodepoolTable, NodepoolRecord, Seq],
+    appQuery: Query[AppTable, AppRecord, Seq],
+    labelFilter: LabelMap = Map()
+  )(implicit ec: ExecutionContext): DBIO[List[KubernetesCluster]] =
+    joinFullApp(baseQuery, nodepoolQuery, appQuery, labelFilter).result
+      .map(rec => aggregateJoinedApp(rec).toList)
+
+  private[db] def joinFullApp(baseClusterQuery: Query[KubernetesClusterTable, KubernetesClusterRecord, Seq],
```

Comment:
the `join`s and `joinLeft`s here work such that even for listing, we will not return any nodepools/clusters with no apps.